doc/man/ldmud.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH LDMUD "1" "July 2025" "LDMud 3.6.8 - a LPMud Game Driver." "User Commands"
.SH NAME
LDMud \- ldmud
.SH SYNOPSIS
.B ldmud
[\fI\,options\/\fR] [\fI\,<portnumber>\/\fR...]
.SH DESCRIPTION
LDMud 3.6.8 \- a LPMud Game Driver.
.PP
Options are:
.HP
\fB\-\-args\fR <filename>
.IP
Read the options from <filename> as if they were given on the
commandline.
.HP
\fB\-P\fR|\-\-inherit <fd\-number>
.IP
Inherit filedescriptor <fd\-number> from the parent process
as socket to listen for connections.
.HP
\fB\-u\fR|\-\-udp <portnumber>
.IP
Specify the <portnumber> for the UDP port, overriding the compiled\-in
default.
.HP
\fB\-D\fR|\-\-define <macro>[=<text>]
.IP
Add <macro> (optionally to be expanded to <text>) to the list of
predefined macros known by the LPC compiler.
.HP
\fB\-E\fR|\-\-eval\-cost <ticks>
.IP
Set the number of <ticks> available for one evaluation thread.
.HP
\fB\-M\fR|\-\-master <filename>
.IP
Use <filename> for the master object.
.HP
\fB\-m\fR|\-\-mudlib <pathname>
.IP
Use <pathname> as the top directory of the mudlib.
.HP
\fB\-\-debug\-file\fR <filename>
.IP
Log all debug output in <filename> instead of <host>.debug.log .
.HP
\fB\-\-access\-file\fR <filename>|none
.IP
Activate access control with these access permissions data.
If 'none' is given access control is deactivated.
.HP
\fB\-\-access\-log\fR <filename>|none
.IP
Log valid and rejected connections into this file.
If 'none' is given no log is written.
.HP
\fB\-\-hostname\fR <name>
.IP
Use <name> as hostname, instead of what the system says.
.HP
\fB\-\-hostaddr\fR <addr>
.IP
Use <addr> as address of this machine, instead of what the
system says. In particular this address will be used to open
the driver ports.
.HP
\fB\-\-compat\fR
.HP
\fB\-\-no\-compat\fR
.IP
Select the mode (compat or plain) of the driver.
Note that this choice does not affect the default name of the master
object.
.HP
\fB\-d\fR|\-\-debug
.IP
Generate debug output; repeat the argument for even more output.
.HP
\fB\-c\fR|\-\-list\-compiles
.IP
List the name of every compiled file on stderr.
.HP
\fB\-e\fR|\-\-no\-preload
.IP
Pass a non\-zero argument (the number of occurrences of this option)
to master\->preload(), which usually inhibits all preloads of castles
and other objects.
.HP
\fB\-\-erq\fR <filename>
.HP
\fB\-\-erq\fR "<filename> <erq arguments>"
.IP
Use <filename> instead of 'erq' as the name of the ERQ executable.
If the name starts with a '/', it is take to be an absolute pathname,
otherwise it is interpreted relative to the executable directory.
If not specified, 'erq' is used as executable name.
With the proper use of quotes it is legal to pass arbitrary arguments
to the erq, however, these may not contain spaces themselves.
.HP
\fB\-N\fR|\-\-no\-erq
.IP
Don't start the erq demon (if it would be started at all).
.HP
\fB\-\-alarm\-time\fR <time>
.IP
The granularity of call_out()s and heartbeats (minimum: 1).
.HP
\fB\-\-heart\-beat\-interval\fR <time>
.IP
The time to elapse between two heartbeats (minimum: 1).
.HP
\fB\-\-synchronous\-heart\-beat\fR
.IP
All heartbeats are executed at the same time (modulo granularity).
.HP
\fB\-\-asynchronous\-hear\-beatt\fR
.IP
Heartbeats are executed immediately when they are due (modulo granularity).
.HP
\fB\-t\fR|\-\-no\-timers
.IP
Disable heartbeats and call_outs.
.HP
\fB\-f\fR|\-\-funcall <word>
.IP
The lfun master\->flag() is called with <word> as argument before the
gamedriver accepts network connections.
.HP
\fB\-\-cleanup\-time\fR <time>
.IP
The idle time in seconds for an object before the driver tries to
clean it up. This time should be substantially higher than the
reset time. A time <= 0 disables the cleanup mechanism.
.HP
\fB\-\-reset\-time\fR <time>
.IP
The time in seconds for an object before it is reset.
A time <= 0 disables the reset mechanism.
.HP
\fB\-\-regexp\fR pcre|traditional
.IP
Select the default regexp package.
.HP
\fB\-\-max\-array\fR <size>
.IP
The maximum number of elements an array can hold.
Set to 0, arrays of any size are allowed.
.HP
\fB\-\-max\-callouts\fR <number>
.IP
The maximum number of callouts at one time.
Set to 0, any number of callouts is allowed.
.HP
\fB\-\-max\-mapping\fR <size>
.IP
The maximum number of elements (keys+values) a mapping can hold.
Set to 0, mappings of any size are allowed.
.HP
\fB\-\-max\-mapping\-keys\fR <size>
.IP
The maximum number of entries (keys) a mapping can hold.
Set to 0, mappings of any size are allowed.
.HP
\fB\-\-max\-bytes\fR <size>
.IP
The maximum number of bytes one read_bytes()/write_bytes() call
can handle.
Set to 0, reads and writes of any size are allowed.
.HP
\fB\-\-max\-file\fR <size>
.IP
The maximum number of bytes one read_file()/write_file() call
can handle.
Set to 0, reads and writes of any size are allowed.
.HP
\fB\-\-max\-write\-buffer\fR <size>
.IP
The maximum number of bytes to be kept pending for each socket
to write.
Set to 0, an unlimited amount of data can be kept pending.
.TP
\fB\-s\fR <time> | \fB\-\-swap\-time\fR <time>
.HP
\fB\-s\fR v<time> | \fB\-\-swap\-variables\fR <time>
.IP
Time in seconds before an object (or its variables) are swapped out.
A time less or equal 0 disables swapping.
.HP
\fB\-s\fR f<name> | \fB\-\-swap\-file\fR <name>
.IP
Swap into file <name> instead of LP_SWAP.<host> .
.HP
\fB\-s\fR c | \fB\-\-swap\-compact\fR
.IP
Reuse free space in the swap file immediately.
.HP
\fB\-\-hard\-malloc\-limit\fR <size>
.IP
Restrict total memory allocation to <size> bytes. A <size> of 0
or 'unlimited' removes any restriction.
.HP
\fB\-\-soft\-malloc\-limit\fR <size>
.IP
If total memory allocation exceeds <size> bytes, inform the mudlib
master about a developing low memory situation. A <size> of 0
or 'unlimited' removes the threshold. <size> must be smaller than
\fB\-\-hard\-malloc\-limit\fR.
.HP
\fB\-\-min\-malloc\fR <size>
.HP
\fB\-\-min\-small\-malloc\fR <size>
.IP
Determine the sizes for the explicit initial large resp. small chunk
allocation. A size of 0 disables the explicit initial allocations.
.HP
\fB\-r\fR u<size> | \fB\-\-reserve\-user\fR <size>
.HP
\fB\-r\fR m<size> | \fB\-\-reserve\-master\fR <size>
.HP
\fB\-r\fR s<size> | \fB\-\-reserve\-system\fR <size>
.IP
Reserve <size> amount of memory for user/master/system allocations to
be held until main memory runs out.
.HP
\fB\-\-filename\-spaces\fR
.HP
\fB\-\-no\-filename\-spaces\fR
.IP
Allow/disallow the use of spaces in filenames.
.HP
\fB\-\-strict\-euids\fR
.HP
\fB\-\-no\-strict\-euids\fR
.IP
Enforce/don't enforce the proper use of euids.
.HP
\fB\-\-share\-variables\fR
.HP
\fB\-\-init\-variables\fR
.IP
Select how clones initialize their variables:
.IP
\- by sharing the current values of their blueprint
.br
\- by initializing them afresh (using __INIT()).
.HP
\fB\-\-tls\-key\fR <pathname>|none
.IP
Use <pathname> as the x509 keyfile, default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
If 'none' is given and no key directory is specified TLS is deactivated.
.HP
\fB\-\-tls\-keydirectory\fR <pathname>|none
.IP
Use <pathname> as the directory where PEM coded files containing
the private key and corresponding x509 certificate reside,
default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
If 'none' is given and no key file is specified TLS is deactivated.
.HP
\fB\-\-tls\-cert\fR <pathname>
.IP
Use <pathname> as the x509 certfile, default is 'cert.pem'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-tls\-trustfile\fR <pathname>|none
.IP
Use <pathname> as the filename holding your trusted PEM certificates,
default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-tls\-trustdirectory\fR <pathname>|none
.IP
Use <pathname> as the directory where your trusted PEM certificates reside,
default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-tls\-crlfile\fR <pathname>|none
.IP
Use <pathname> as the filename holding your certificate revocation lists,
default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-tls\-crldirectory\fR <pathname>|none
.IP
Use <pathname> as the directory where your certificate revocation lists reside,
default is 'none'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-tls\-password\fR pass:<password>|env:<variable>|file:<pathname>|fd:<number>|stdin|none
.IP
Decode the TLS keyfile with the given password:
.RS
.TP
\- pass:<password>
Use the password given on the command line.
.TP
\- env:<variable>
Read the password from the environment variable.
.TP
\- file:<pathname>
Read the password from the given file.
.TP
\- fd:<number>
Read the password from the file descriptor.
.TP
\- stdin
Prompt for the password on standard input.
.TP
\- none
There is no password.
.RE
.IP
default is 'none'.
.IP
The password is read at program startup and upon calls to tls_refresh_certs().
If relative, <pathname> is interpreted relative to <mudlib>. Only the first line
will be read from files (file, fd and stdin).
.HP
\fB\-\-python\-script\fR <pathname>
.IP
Execute the python script <pathname> on startup, default is 'startup.py'.
If relative, <pathname> is interpreted relative to <mudlib>.
.HP
\fB\-\-wizlist\-file\fR <filename>
.HP
\fB\-\-no\-wizlist\-file\fR
.IP
Read and save the wizlist in the named file (always interpreted
relative the mudlib); resp. don't read or save the wizlist.
.HP
\fB\-\-pidfile\fR <filename>
.IP
Write the pid of the driver process into <filename>.
.HP
\fB\-\-randomdevice\fR <filename>
.IP
Determines the source of the seed for the random number generator.
(tries \fI\,/dev/urandom\/\fP by default and uses system clock as fallback)
.HP
\fB\-\-random\-seed\fR <num>
.IP
Seed value for the random number generator. If not given, the
driver chooses a seed value on its own.
.HP
\fB\-\-gcollect\-outfd\fR <filename>|<num>
.IP
Garbage collector output (like a log of all reclaimed memory blocks)
is sent to <filename> (or inherited fd <num>) instead of stderr.
.HP
\fB\-\-check\-refcounts\fR
.IP
Every backend cycle, all refcounts in the system are checked.
SLOW!
.HP
\fB\-\-check\-state\fR <lvl>
.IP
Perform a regular simplistic check of the virtual machine according
to <lvl>:
.IP
= 0: no check
.br
= 1: once per backend loop
.br
= 2: at various points in the backend loop
.HP
\fB\-\-gobble\-descriptors\fR <num>
.IP
<num> (more) filedescriptors are used up. You'll know when you need it.
.HP
\fB\-\-options\fR
.IP
Print the version and compilation options of the driver, then exit.
.HP
\fB\-V\fR|\-\-version
.IP
Print the version of the driver, then exit.
.HP
\fB\-\-longhelp\fR
.IP
Display this help and exit.
.HP
\fB\-h\fR|\-?|\-\-help
.IP
Display the short help text and exit.
.SH AUTHOR
Maintained by the LDMud team (ldmud-dev@unitopia.de).
.SH "REPORTING BUGS"
Bug reports and feature suggestions are tracked online at
https://mantis.ldmud.eu/mantis/.
.SH COPYRIGHT
Copyright 1997-2007 by Lars Düning.
.br
Copyright 2008-2025 by Bertram Felgenhauer, Alexander Motzkau and Dominik Schäfer.

LDMud is based on LPmud, which is
.br
Copyright 1989-1991 by Lars Pensjö, and, starting with version 3.1.2,
.br
Copyright 1991-1997 by Jörn Rennecke.

Source code herein refers to the source code, and any executables
created from the same source code.

All rights reserved. Permission is granted to use, extend and modify the
source code provided subject to the restriction that the source code may
not be used in any way whatsoever for monetary gain.

LDMud contains additional code which are copyright by their respective
authors - see the files COPYRIGHT and CREDITS for details.


CHANGES
Changes from 1.4.0-A to 1.4.1-A

Fixed several more bugs.  These constitute the majority of changes.  Among
others, changes were made to the wizard shutdown and time commands, the
player describe command, the initial loading routine, the error handling
routine, and monster.c.  (This list is far from complete.)

Upon an error, players are now moved to the church instead of to the void,
assuming the game is able to do this.  Everything else still goes to the
void.

The ability to put traps on doors was added.  See /doc/build/door for details.

The ability to use "stat" has been changed (in the version of security.h
included here) to allow anyone of CREATE level to use it.  However, living.h
has been changed so that unless the wizard is of at least STAT level, he
will not see the monster's current hp, so he can't tip off players.


gdb/ldmud.py
import gdb

def unwind_ptr(val):
    type = val.type

    while True:
        if type.code == gdb.TYPE_CODE_REF:
            type = type.target()
        elif type.code == gdb.TYPE_CODE_PTR:
            val = val.dereference()
            type = val.type
        elif type.code == gdb.TYPE_CODE_ARRAY:
            val = val[0]
            type = val.type
        else:
            break

    return (type, val)

def print_ptr(val):
    return "0x%x" % (val,)

class MStringPrinter:
    "Print an mstring"

    STRING_ASCII = 0
    STRING_UTF8  = 1
    STRING_BYTES = 2

    def __init__(self, val):
        self.val = val

    def to_string(self):
        # Handle references
        (type,val) = unwind_ptr(self.val)

        if val.address == 0:
            return "0x0"

        if val["info"]["unicode"] == self.STRING_BYTES:
            result = '"'
            for idx in range(val["size"]):
                ch = val["txt"][idx]
                if ch < 0:
                    ch += 256
                result += '\\x%02x' % (ch,)
            result += '"'
            return result
        else:
            # Some GDB versions have a broken implementation
            # of gdb.Value.string(), therefore doing manually
            # val["txt"].string('utf-8', length = val["size"])
            b = []
            for idx in range(val["size"]):
                ch = val["txt"][idx]
                if ch < 0:
                    ch += 256
                b.append(ch)
            return '"' + bytes(b).decode('utf-8') + '"'

class PtrNamePrinter:
    "Print an pointer to a struct with an name entry"

    def __init__(self, val, indices):
        self.val = val
        self.indices = indices

    def to_string(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return "0x0"

        nameval = val;
        for idx in self.indices:
            (type, nameval) = unwind_ptr(nameval[idx])
            if nameval.address == 0:
                return print_ptr(val.address)

        return print_ptr(val.address) + ' ' + MStringPrinter(nameval).to_string()

class ArrayPrinter:
    "Print an LPC array"

    class Iterator:
        def __init__(self, items, size):
            self.items = items
            self.size = int(size)
            self.pos = 0

        def __iter__(self):
            return self

        def __next__(self):
            if self.pos >= self.size:
                raise StopIteration

            result = ('[%d]' % (self.pos), self.items[self.pos].address)

            self.pos += 1

            return result

        def next(self):
            return self.__next__()

    def __init__(self, val):
        self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return "0x0"

        return print_ptr(val.address) + " Array with %d element%s" % (val["size"], val["size"]!=1 and "s" or "")

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return self.Iterator(self.val["item"], self.val["size"])

    def display_hint(self):
        return "array"

class MappingPrinter:
    "Print an LPC mapping"

    class Iterator:
        def __init__(self, mapping):
            self.mapping = mapping
            self.width = mapping["num_values"]+1
            self.size = mapping["num_entries"]
            if mapping["cond"] == 0:
                self.condsize = 0
            else:
                self.condpart = mapping["cond"].dereference()
                self.condsize = self.condpart["size"]

            if mapping["hash"] == 0:
                self.hashsize = 0
            else:
                self.hashpart = mapping["hash"].dereference()
                self.hashsize = self.hashpart["used"]
                self.hashchains = self.hashpart["mask"] + 1

            self.pos = 0
            self.hashchain = 0
            self.hashpos = 0

        def __iter__(self):
            return self

        def __next__(self):
            item = None

            key = self.pos / self.width
            val = self.pos % self.width

            if key < self.condsize:
                # We have to look in the condensed part
                if val == 0:
                    item = self.condpart["data"][key].address
                else:
                    item = self.condpart["data"][self.condsize + key * (self.width-1) + val-1].address
            elif key < self.condsize + self.hashsize:
                while self.hashpos < self.hashsize and not self.hashchain:
                    self.hashchain = self.hashpart["chains"][self.hashpos]
                    self.hashpos += 1
                if self.hashchain:
                    item = self.hashchain.dereference()["data"][val].address

            if item == None:
                raise StopIteration

            if val == 0:
                result = ('[%d]' % (key), item)
            else:
                result = ('[%d,%d]' % (key, val-1), item)

            self.pos += 1
            if self.hashchain and (self.pos % self.width) == 0:
                self.hashchain = self.hashchain.dereference()["next"]

            return result

        def next(self):
            return self.__next__()

    def __init__(self, val):
        self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return "0x0"

        return print_ptr(val.address) + " Mapping with %d %s and %d value%s" % (val["num_entries"], val["num_entries"]!=1 and "entries" or "entry", val["num_values"], val["num_values"]!=1 and "s" or "")

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return self.Iterator(self.val)

#    def display_hint(self):
#        return "map"

class SValuePrinter:
    "Print an svalue"

    T_INVALID                        = 0x0
    T_LVALUE                         = 0x1
    T_NUMBER                         = 0x2
    T_STRING                         = 0x3
    T_POINTER                        = 0x4
    T_OBJECT                         = 0x5
    T_MAPPING                        = 0x6
    T_FLOAT                          = 0x7
    T_CLOSURE                        = 0x8
    T_SYMBOL                         = 0x9
    T_QUOTED_ARRAY                   = 0xa
    T_STRUCT                         = 0xb
    T_BYTES                          = 0xc
    T_LWOBJECT                       = 0xd
    T_COROUTINE                      = 0xe
    T_PYTHON                         = 0xf
    T_LPCTYPE                        = 0x10
    T_CALLBACK                       = 0x11
    T_ERROR_HANDLER                  = 0x12
    T_BREAK_ADDR                     = 0x13
    T_ARG_FRAME                      = 0x14
    T_NULL                           = 0x15

    LVALUE_UNPROTECTED               = 0x00
    LVALUE_UNPROTECTED_CHAR          = 0x01
    LVALUE_UNPROTECTED_RANGE         = 0x02
    LVALUE_UNPROTECTED_MAPENTRY      = 0x03
    LVALUE_UNPROTECTED_MAP_RANGE     = 0x04
    LVALUE_PROTECTED                 = 0x10
    LVALUE_PROTECTED_CHAR            = 0x11
    LVALUE_PROTECTED_RANGE           = 0x12
    LVALUE_PROTECTED_MAPENTRY        = 0x13
    LVALUE_PROTECTED_MAP_RANGE       = 0x14

    names = {
        T_INVALID:                        "T_INVALID",
        T_LVALUE:                         "T_LVALUE",
        T_NUMBER:                         "T_NUMBER",
        T_STRING:                         "T_STRING",
        T_POINTER:                        "T_POINTER",
        T_OBJECT:                         "T_OBJECT",
        T_MAPPING:                        "T_MAPPING",
        T_FLOAT:                          "T_FLOAT",
        T_CLOSURE:                        "T_CLOSURE",
        T_SYMBOL:                         "T_SYMBOL",
        T_QUOTED_ARRAY:                   "T_QUOTED_ARRAY",
        T_STRUCT:                         "T_STRUCT",
        T_BYTES:                          "T_BYTES",
        T_LWOBJECT:                       "T_LWOBJECT",
        T_COROUTINE:                      "T_COROUTINE",
        T_PYTHON:                         "T_PYTHON",
        T_LPCTYPE:                        "T_LPCTYPE",
        T_CALLBACK:                       "T_CALLBACK",
        T_ERROR_HANDLER:                  "T_ERROR_HANDLER",
        T_BREAK_ADDR:                     "T_BREAK_ADDR",
        T_ARG_FRAME:                      "T_ARG_FRAME",
        T_NULL:                           "T_NULL",
    }

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return "0x0"

        stype = int(val["type"])
        if stype in self.names:
            return print_ptr(val.address) + " " + self.names[stype]

        return print_ptr(val.address) + " svalue_t with unknown type %d" % (stype)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        stype = val["type"]
        if stype == self.T_LVALUE:
            ltype = val["x"]["lvalue_type"]

            if ltype == self.LVALUE_UNPROTECTED:
                return [("x.lvalue_type", "LVALUE_UNPROTECTED"),
                        ("u.lvalue", val["u"]["lvalue"])]
            elif ltype == self.LVALUE_UNPROTECTED_CHAR:
                lvalue = gdb.lookup_global_symbol("current_unprotected_char").value()
                return [("x.lvalue_type", "LVALUE_UNPROTECTED_CHAR"),
                        ("current_unprotected_char", lvalue)]
            elif ltype == self.LVALUE_UNPROTECTED_RANGE:
                lvalue = gdb.lookup_global_symbol("current_unprotected_range").value()
                return [("x.lvalue_type", "LVALUE_UNPROTECTED_RANGE"),
                        ("current_unprotected_range", lvalue)]
            elif ltype == self.LVALUE_UNPROTECTED_MAPENTRY:
                lvalue = gdb.lookup_global_symbol("current_unprotected_mapentry").value()
                return [("x.lvalue_type", "LVALUE_UNPROTECTED_MAPENTRY"),
                        ("current_unprotected_mapentry", lvalue)]
            elif ltype == self.LVALUE_UNPROTECTED_MAP_RANGE:
                lvalue = gdb.lookup_global_symbol("current_unprotected_map_range").value()
                return [("x.lvalue_type", "LVALUE_UNPROTECTED_MAP_RANGE"),
                        ("current_unprotected_map_range", lvalue)]
            elif ltype == self.LVALUE_PROTECTED:
                return [("x.lvalue_type", "LVALUE_PROTECTED"),
                        ("u.protected_lvalue", val["u"]["protected_lvalue"])]
            elif ltype == self.LVALUE_PROTECTED_CHAR:
                return [("x.lvalue_type", "LVALUE_PROTECTED_CHAR"),
                        ("u.protected_char_lvalue", val["u"]["protected_char_lvalue"])]
            elif ltype == self.LVALUE_PROTECTED_RANGE:
                return [("x.lvalue_type", "LVALUE_PROTECTED_RANGE"),
                        ("u.protected_range_lvalue", val["u"]["protected_range_lvalue"])]
            elif ltype == self.LVALUE_PROTECTED_MAPENTRY:
                return [("x.lvalue_type", "LVALUE_PROTECTED_MAP_ENTRY"),
                        ("u.protected_mapentry_lvalue", val["u"]["protected_mapentry_lvalue"])]
            elif ltype == self.LVALUE_PROTECTED_MAP_RANGE:
                return [("x.lvalue_type", "LVALUE_PROTECTED_MAPRANGE"),
                        ("u.protected_map_range_lvalue", val["u"]["protected_map_range_lvalue"])]
            else:
                return [(".x.lvalue_type", ltype)]
        elif stype == self.T_NUMBER:
            return [(".u.number", val["u"]["number"])]
        elif stype == self.T_STRING or stype == self.T_BYTES:
            return [(".u.str", val["u"]["str"])]
        elif stype == self.T_POINTER:
            return [(".u.vec", val["u"]["vec"])]
        elif stype == self.T_OBJECT:
            return [(".u.ob", val["u"]["ob"])]
        elif stype == self.T_LWOBJECT:
            return [(".u.lwob", val["u"]["lwob"])]
        elif stype == self.T_MAPPING:
            return [(".u.map", val["u"]["map"])]
        elif stype == self.T_FLOAT:
            if any([field.name == "float_number" for field in val["u"].type.fields()]):
                return [("u.float_number", val["u"]["float_number"])]
            return [(".u.mantissa", val["u"]["mantissa"]),
                    (".x.exponent", val["x"]["exponent"])]
        elif stype == self.T_CLOSURE:
            return [(".u.lambda", val["u"]["lambda"]),
                    (".x.closure_type", val["x"]["closure_type"])]
        elif stype == self.T_COROUTINE:
            return [(".u.coroutine", val["u"]["coroutine"])]
        elif stype == self.T_SYMBOL:
            return [(".u.str", val["u"]["str"]),
                    (".x.quotes", val["x"]["quotes"])]
        elif stype == self.T_QUOTED_ARRAY:
            return [(".u.vec", val["u"]["vec"]),
                    (".x.quotes", val["x"]["quotes"])]
        elif stype == self.T_STRUCT:
            return [(".u.strct", val["u"]["strct"])]
        elif stype == self.T_PYTHON:
            return [(".u.generic", val["u"]["generic"])]
        elif stype == self.T_LPCTYPE:
            return [(".u.lpctype", val["u"]["lpctype"])]
        elif stype == self.T_CALLBACK:
            return [(".u.cb", val["u"]["cb"])]
        elif stype == self.T_ERROR_HANDLER:
            return [(".u.error_handler", val["u"]["error_handler"])]
        elif stype == self.T_BREAK_ADDR:
            return [(".u.break_addr", val["u"]["break_addr"])]
        elif stype == self.T_ARG_FRAME:
            return [(".u.lvalue", val["u"]["lvalue"])]
        else:
            return []

class ProtLvalPrinter:
    "Prints a protected lvalue"

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)

        return print_ptr(val.address)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return [(".ref", val["ref"]),
                (".val", val["val"].address)]

class ProtCharPrinter:
    "Prints a protected char lvalue"

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)

        return print_ptr(val.address)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return [(".ref", val["ref"]),
                (".str", val["str"]),
                (".charp", val["charp"])]

class ProtRangePrinter:
    "Prints a protected range lvalue"

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)

        return print_ptr(val.address)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return [(".ref", val["ref"]),
                (".index1", val["index1"]),
                (".index2", val["index2"]),
                (".vec", val["vec"].address),
                (".var", val["var"])]

class ProtMapEntryPrinter:
    "Prints a protected map entry lvalue"

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)

        return print_ptr(val.address)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return [(".ref", val["ref"]),
                (".map", val["map"]),
                (".key", val["key"].address),
                (".index", val["index"])]

class ProtMapRangePrinter:
    "Prints a protected map range lvalue"

    def __init__(self, val):
         self.val = val

    def to_string(self):
        (type,val) = unwind_ptr(self.val)

        return print_ptr(val.address)

    def children(self):
        (type,val) = unwind_ptr(self.val)
        if val.address == 0:
            return []

        return [(".ref", val["ref"]),
                (".map", val["map"]),
                (".key", val["key"].address),
                (".index1", val["index1"]),
                (".index2", val["index2"])]

class TypePrinter:
    "Prints a LPC type"

    TCLASS_PRIMARY    = 0
    TCLASS_STRUCT     = 1
    TCLASS_OBJECT     = 2
    TCLASS_ARRAY      = 3
    TCLASS_UNION      = 4
    TCLASS_PYTHON     = 5

    TYPE_UNKNOWN      =  0
    TYPE_NUMBER       =  1
    TYPE_STRING       =  2
    TYPE_VOID         =  3
    TYPE_MAPPING      =  4
    TYPE_FLOAT        =  5
    TYPE_ANY          =  6
    TYPE_CLOSURE      =  7
    TYPE_SYMBOL       =  8
    TYPE_QUOTED_ARRAY =  9
    TYPE_BYTES        = 10
    TYPE_COROUTINE    = 11
    TYPE_LPCTYPE      = 12

    OBJECT_REGULAR    =  0
    OBJECT_LIGHTWEIGHT=  1

    type_names = {
        TYPE_UNKNOWN:      "unknown",
        TYPE_NUMBER:       "int",
        TYPE_STRING:       "string",
        TYPE_VOID:         "void",
        TYPE_MAPPING:      "mapping",
        TYPE_FLOAT:        "float",
        TYPE_ANY:          "mixed",
        TYPE_CLOSURE:      "closure",
        TYPE_SYMBOL:       "symbol",
        TYPE_QUOTED_ARRAY: "quoted_array",
        TYPE_BYTES:        "bytes",
        TYPE_COROUTINE:    "coroutine",
        TYPE_LPCTYPE:      "lpctype",
    }

    def __init__(self, val):
        self.val = val

    def calc_name(self, val):
        if val.address == 0:
            return "0x0"

        tclass = int(val["t_class"])

        if tclass == self.TCLASS_PRIMARY:
            primary = int(val["t_primary"])
            if primary in self.type_names:
                return self.type_names[primary]
            return "T_PRIMARY with unknown type %d" % (primary,)

        elif tclass == self.TCLASS_STRUCT:
            structinfo = val["t_struct"]["name"].dereference()
            if structinfo.address == 0:
                return "any struct"

            progname = structinfo["prog_name"]
            basename = structinfo["name"]

            return "struct %s:%s" % (
                progname["txt"].string(length = progname["size"]),
                basename["txt"].string(length = basename["size"]))

        elif tclass == self.TCLASS_OBJECT:
            progname = val["t_object"]["program_name"].dereference()
            if int(val["t_object"]["type"]) == self.OBJECT_LIGHTWEIGHT:
                lw = "lightweight "
            else:
                lw = ""

            if progname.address == 0:
                return "any %sobject" % (lw,)

            return '%sobject "%s"' % (lw, progname["txt"].string(length = progname["size"]))

        elif tclass == self.TCLASS_ARRAY:
            return self.calc_name(val["t_array"]["base"].dereference()) + "*"*int(val["t_array"]["depth"])

        elif tclass == self.TCLASS_UNION:
            typestr = ""
            union = val
            while tclass == self.TCLASS_UNION:
                typestr += self.calc_name(union["t_union"]["member"].dereference()) + "|"
                union = union["t_union"]["head"].dereference()
                tclass = int(union["t_class"])
            typestr += self.calc_name(union)
            return "<" + typestr + ">"

        elif tclass == self.TCLASS_PYTHON:
            return "python type %d" % (int(val["t_python"]["type_id"]),)

        else:
            return "lpctype_t with unknown class %d" % (tclass,)

    def to_string(self):
        (type,val) = unwind_ptr(self.val)
        return print_ptr(val.address) + " " + self.calc_name(val)

def lookup_printer(val):
    "Returns a pretty printer for val (gdb.Value)."

    # Handle references and pointers
    type = val.type
    haveptr = False
    while type.code == gdb.TYPE_CODE_REF or ((type.code == gdb.TYPE_CODE_PTR or type.code == gdb.TYPE_CODE_ARRAY) and not haveptr):
        if type.code == gdb.TYPE_CODE_PTR or type.code == gdb.TYPE_CODE_ARRAY:
            haveptr = True
        type = type.target()

    type = type.strip_typedefs()

    if type.tag == None:
        return None

    if haveptr:
        printerlist = ptr_printers
    else:
        printerlist = direct_printers

    if type.tag in printerlist:
        return printerlist[type.tag](val)

    return None

def register_printers(obj = None):
    "Register pretty printers for LDMud with obj."

    if obj == None:
        obj = gdb

    obj.pretty_printers.append(lookup_printer)

ptr_printers = {
    'string_s':      MStringPrinter,
    'vector_s':      ArrayPrinter,
    'mapping_s':     MappingPrinter,
    'svalue_s':      SValuePrinter,
    'object_s':      lambda val: PtrNamePrinter(val, ["name"]),
    'lwobject_s':    lambda val: PtrNamePrinter(val, ["prog", "name"]),
    'program_s':     lambda val: PtrNamePrinter(val, ["name"]),
    'interactive_s': lambda val: PtrNamePrinter(val, ["ob", "name"]),
    'wiz_list_s':    lambda val: PtrNamePrinter(val, ["name"]),
    'lpctype_s':     TypePrinter,
    'protected_lvalue':           ProtLvalPrinter,
    'protected_char_lvalue':      ProtCharPrinter,
    'protected_range_lvalue':     ProtRangePrinter,
    'protected_mapentry_lvalue':  ProtMapEntryPrinter,
    'protected_map_range_lvalue': ProtMapRangePrinter,
}

direct_printers = {
    'svalue_s':      SValuePrinter,
}
